### PR TITLE
OutputControl:Files

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

 - Add new `OutputControl:Files` object to SDK

Please read [OpenStudio Pull Requests](https://github.com/NREL/OpenStudio/wiki/OpenStudio-Pull-Requests) to better understand the OpenStudio Pull Request protocol.

### Pull Request Author

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Checked behavior in OpenStudioApplication, adjusted policies as needed (`src/openstudio_lib/library/OpenStudioPolicy.xml`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

**Comment by tijcolem** [comment]
Nice work on this @joseph-robertson.  Just a note that CI is only set up to run on PRs to develop branch so it won't automatically run on this PR.  I kicked off a full build for this branch instead.    https://ci.commercialbuildings.dev/blue/organizations/jenkins/openstudio-develop-nightly/detail/openstudio-develop-nightly/685/pipeline

File: src/model/test/OutputControlFiles_GTest.cpp [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+  EXPECT_FALSE(outputControlFiles.outputExtShd());
+  EXPECT_TRUE(outputControlFiles.setOutputExtShd(true));
+  EXPECT_TRUE(outputControlFiles.outputExtShd());
+
+  EXPECT_TRUE(outputControlFiles.outputTarcog());
+  EXPECT_TRUE(outputControlFiles.setOutputTarcog(false));
+  EXPECT_FALSE(outputControlFiles.outputTarcog());
+  EXPECT_TRUE(outputControlFiles.setOutputTarcog(true));
+  EXPECT_TRUE(outputControlFiles.outputTarcog());
+}
```
Comment (jmarrec) [inline src/model/test/OutputControlFiles_GTest.cpp:228]: Straightforward Model Gtest, no problem.

File: src/model/test/UniqueModelObject_GTest.cpp [hunk 2]
```diff
@@ -347,6 +349,32 @@ TEST_F(ModelFixture, LightingSimulationControl_UniqueModelObject_Clone)
...
+  EXPECT_EQ("! Custom Object", outputControlFilesClone.comment());
+
+  // clone it into a different model
+  Model model2;
+  EXPECT_FALSE(model2.getOptionalUniqueModelObject<OutputControlFiles>());
+  OutputControlFiles outputControlFilesClone2 = outputControlFiles.clone(model2).cast<OutputControlFiles>();
+  EXPECT_TRUE(model2.getOptionalUniqueModelObject<OutputControlFiles>());
+  EXPECT_EQ("! Custom Object", outputControlFilesClone2.comment());
+}
+
```
Comment (jmarrec) [inline src/model/test/UniqueModelObject_GTest.cpp:377]: Nice to have thought of adding it here too! :+1:

File: src/model/OutputControlFiles.cpp [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+                                                   Model_Impl* model,
+                                                   bool keepHandle)
+    : ModelObject_Impl(other,model,keepHandle)
+  {}
+
+  const std::vector<std::string>& OutputControlFiles_Impl::outputVariableNames() const
+  {
+    static const std::vector<std::string> result;
+    if (result.empty()){
+    }
```
Comment (jmarrec) [inline src/model/OutputControlFiles.cpp:73]: Remove these two lines.

File: src/model/OutputControlFiles_Impl.hpp [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+                            bool keepHandle);
+
+    virtual ~OutputControlFiles_Impl() {}
+
+    //@}
+
+    /** @name Virtual Methods */
+    //@{
+
+    virtual const std::vector<std::string>& outputVariableNames() const override;
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/OutputControlFiles_Impl.hpp:67]: Nevermind, I forgot ModelObject_Impl::outputVariableNames was a pure virtual method (`= 0`)

File: src/model/OutputControlFiles.cpp [hunk 5]
```diff
@@ -0,0 +1,628 @@
...
+  setOutputDelightDFdmp(true);
+  setOutputEDD(true);
+  setOutputDBG(true);
+  setOutputPerfLog(true);
+  setOutputSLN(true);
+  setOutputSCI(true);
+  setOutputWRL(true);
+  setOutputScreen(true);
+  setOutputExtShd(true);
+  setOutputTarcog(true);
```
Comment (jmarrec) [inline src/model/OutputControlFiles.cpp:622]: Ok, this matches the defaults in the IDD: all true, except the Output MTR (for some reason).

File: src/model/OutputControlFiles.cpp [hunk 6]
```diff
@@ -0,0 +1,628 @@
...
+  return getImpl<detail::OutputControlFiles_Impl>()->setOutputTarcog(outputTarcog);
+}
+
+/// @cond
+OutputControlFiles::OutputControlFiles(std::shared_ptr<detail::OutputControlFiles_Impl> impl)
+  : ModelObject(std::move(impl))
+{}
+OutputControlFiles::OutputControlFiles(Model& model)
+  : ModelObject(OutputControlFiles::iddObjectType(),model)
+{
```
Comment (jmarrec) [inline src/model/OutputControlFiles.cpp:589]: Really a cosmetic thing, but I prefer when the Ctors are at the begining of the public declarations (right after the end of the detail (Impl) namespace block). That's where I go to look for them usually. I don't mind though.

File: src/model/OutputControlFiles.cpp [hunk 7]
```diff
@@ -0,0 +1,628 @@
...
+
+} // detail
+
+IddObjectType OutputControlFiles::iddObjectType() {
+  return IddObjectType(IddObjectType::OS_OutputControl_Files);
+}
+
+bool OutputControlFiles::outputCSV() const {
+  return getImpl<detail::OutputControlFiles_Impl>()->outputCSV();
+}
```
Comment (jmarrec) [inline src/model/OutputControlFiles.cpp:337]: All public forward starting here and until end are fine

File: src/model/OutputControlFiles.cpp (hunk 3 above)
Comment (jmarrec) [reply src/model/OutputControlFiles.cpp:73]:
```suggestion
```

File: src/model/OutputControlFiles.cpp [hunk 8]
```diff
@@ -0,0 +1,628 @@
...
+    return result;
+  }
+
+  IddObjectType OutputControlFiles_Impl::iddObjectType() const {
+    return OutputControlFiles::iddObjectType();
+  }
+
+  bool OutputControlFiles_Impl::outputCSV() const {
+    return getBooleanFieldValue(OS_OutputControl_FilesFields::OutputCSV);
+  }
```
Comment (jmarrec) [inline src/model/OutputControlFiles.cpp:83]: Ok Impl side is fine as well, straightforward and without typos.

File: src/energyplus/ForwardTranslator/ForwardTranslateOutputControlFiles.cpp [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+boost::optional<IdfObject> ForwardTranslator::translateOutputControlFiles( OutputControlFiles& modelObject )
+{
+  IdfObject idfObject( openstudio::IddObjectType::OutputControl_Files );
+  m_idfObjects.push_back(idfObject);
+
+  if (modelObject.outputCSV()) {
+    idfObject.setString(OutputControl_FilesFields::OutputCSV, "Yes");
+  } else {
+    idfObject.setString(OutputControl_FilesFields::OutputCSV, "No");
+  }
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslateOutputControlFiles.cpp:54]: (Just a comment: this is perfectly fine. Yet I know how many of these I'm about to read, which makes me wonder why we do not have a high level setBool / setChoice method to facilitate this already.)

File: src/energyplus/ForwardTranslator/ForwardTranslateOutputControlFiles.cpp [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+    idfObject.setString(OutputControl_FilesFields::OutputExtShd, "Yes");
+  } else {
+    idfObject.setString(OutputControl_FilesFields::OutputExtShd, "No");
+  }
+
+  if (modelObject.outputTarcog()) {
+    idfObject.setString(OutputControl_FilesFields::OutputTarcog, "Yes");
+  } else {
+    idfObject.setString(OutputControl_FilesFields::OutputTarcog, "No");
+  }
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslateOutputControlFiles.cpp:234]: FT is all good, all 31 are there and without typos.

File: src/energyplus/ReverseTranslator/ReverseTranslateOutputControlFiles.cpp [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+  if (boost::optional<std::string> _outputTarcog = workspaceObject.getString(OutputControl_FilesFields::OutputTarcog, true)) {
+    if(istringEqual("Yes", _outputTarcog.get())) {
+      modelObject.setOutputTarcog(true);
+    } else {
+      modelObject.setOutputTarcog(false);
+    }
+  }
+
+  result = modelObject;
+  return result;
```
Comment (jmarrec) [inline src/energyplus/ReverseTranslator/ReverseTranslateOutputControlFiles.cpp:300]: Ok, all good, all 31 in there.

File: src/energyplus/Test/OutputControlFiles_GTest.cpp [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+  Workspace w = ft.translateModel(m);
+
+  WorkspaceObjectVector idfObjs = w.getObjectsByType(IddObjectType::OutputControl_Files);
+  ASSERT_EQ(1u, idfObjs.size());
+
+  WorkspaceObject idf_outputControlFiles(idfObjs[0]);
+
+  EXPECT_EQ("No", idf_outputControlFiles.getString(OutputControl_FilesFields::OutputCSV).get());
+  EXPECT_EQ("No", idf_outputControlFiles.getString(OutputControl_FilesFields::OutputMTR).get());
```
Comment (jmarrec) [inline src/energyplus/Test/OutputControlFiles_GTest.cpp:70]: I would suggest doing something similar to what I've done for OutputJSON (test all cases where a single one is true, to make absolutely sure we assigned the fields correctly). Using scripting to write the C++ it's not that annoying.

**Review by jmarrec — APPROVED** [review]
Nice & polished PR, great work @joseph-robertson 


I'll modify the FT/RT tests quickly

The cache-enabled getter in Model is something we should consider, I'm not sure it's really needed or not.

File: src/energyplus/Test/OutputControlFiles_GTest.cpp (hunk 12 above)
Comment (jmarrec) [reply src/energyplus/Test/OutputControlFiles_GTest.cpp:70]: Done

File: src/model/Model.cpp [hunk 13]
```diff
@@ -3506,6 +3507,7 @@ detail::Model_Impl::ModelObjectCreator::ModelObjectCreator() {
   REGISTER_COPYCONSTRUCTORS(Node);
   REGISTER_COPYCONSTRUCTORS(OtherEquipment);
   REGISTER_COPYCONSTRUCTORS(OtherEquipmentDefinition);
+  REGISTER_COPYCONSTRUCTORS(OutputControlFiles);
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/Model.cpp:3545]: I added it in https://github.com/NREL/OpenStudio/pull/4059/commits/2574c9855642bf29cb156906f01349febfc91857, we can revert that if someone objects (@kbenne, @joseph-robertson, @tijcolem ?)

**Comment by jmarrec** [comment]
I rebased onto latest v940-IOFreeze and force-pushed, make sure you refresh your index (git fetch --all)

**Comment by jmarrec** [comment]
If you're comfortable with the tiny changes I've made, go ahead and merge that to v940-IOFreeze!

File: src/model/OutputControlFiles.cpp (hunk 5 above)
Comment (joseph-robertson) [reply src/model/OutputControlFiles.cpp:620]: MTR and CSV

**Pull request merged**